docs/docs/design/grandpa-client.md

```diff
@@ -0,0 +1,272 @@
+	/// Notifies the sync service to try and sync the given block from the given
+	/// peers.
+	///
+	/// If the given vector of peers is empty then the underlying implementation
+	/// should make a best effort to fetch the block from any peers it is
+	/// connected to (NOTE: this assumption will change in the future #3629).
+	fn set_sync_fork_request(&self, peers: Vec<PeerId>, hash: BlockHash, number: BlockNumber);
+}
+
+/// Provides ability to announce blocks to the network.
+pub trait NetworkBlock<BlockHash, BlockNumber> {
+	/// Make sure an important block is propagated to peers.
+	///
+	/// In chain-based consensus, we often need to make sure non-best forks are
+	/// at least temporarily synced. This function forces such an announcement.
+	fn announce_block(&self, hash: BlockHash, data: Option<Vec<u8>>);
+
+	/// Inform the network service about new best imported block.
+	fn new_best_block_imported(&self, hash: BlockHash, number: BlockNumber);
+}
+
+pub trait SyncEventStream: Send + Sync {
+	/// Subscribe to syncing-related events.
+	fn event_stream(&self, name: &'static str) -> Pin<Box<dyn Stream<Item = SyncEvent> + Send>>;
+}
+```
+
+`set_sync_fork_request` is called by [`GossipEngine`].  `announce_block` is also called through `GossipEngine.announce()` via `OutgoingMessages` sink type in `communication` crate.  Doesn't look like `new_best_block_imported` is actually used. `event_stream` is used in [`GossipEngine`] constructor.
+
+### Implementation of `BlockImport`
+
+There is a [`BlockImport`](https://github.com/paritytech/polkadot-sdk/blob/030cb4a71b0b390626a586bfe7117b7c66b4700c/substrate/client/consensus/common/src/block_import.rs#L308) trait that `sc_consensus_grandpa` implements.  In substrate, the `BlockImport` trait is called through a pipelining type called [`BasicQueue`].  An instance of [`BasicQueue`] creates a pipeline of calls to a sequence of implementations of `BlockImport`.  [`BlockImportParams`](https://github.com/paritytech/polkadot-sdk/blob/030cb4a71b0b390626a586bfe7117b7c66b4700c/substrate/client/consensus/common/src/block_import.rs#L170) are provided to each implementation as it progresses through the pipeline for `BlockImport::import_block` function, and is expected to return an `ImportResult` which an enum where the [`ImportResult::Imported`](https://github.com/paritytech/polkadot-sdk/blob/030cb4a71b0b390626a586bfe7117b7c66b4700c/substrate/client/consensus/common/src/block_import.rs#L34) variant is of type [`ImportedAux`](https://github.com/paritytech/polkadot-sdk/blob/030cb4a71b0b390626a586bfe7117b7c66b4700c/substrate/client/consensus/common/src/block_import.rs#L47), which contains data associated with the imported block.  Given that this is the process in substrate, both BEEFY and GRANDPA client implementations currently implement this, and the final implementation in the pipeline is `Client`, I think it makes sense to replicate this functionality given we already already begun replicating the `Client` type.
```

Comment:
You are proposing we fully refactor how gossamer imports blocks? This would quite a big refactor right?

Reply:
Not exactly.  Currently we have design work and issues for implementing `BlockState.AddBlock` functionality #4466. By the time we implement the new version of GRANDPA, we should refactor this translation method `AddBlock` method using the new implementation of `BlockImport` interface for both grandpa client and for `Client`.  High level the logic currently exists in sync, and in BABE to do these finality related checks and additions to the block. I'm propsing we will be move the domain specific logic to their appropriate spots in the pipeline, and removes the calling of said logic in BABE and in sync.

Reply:
Added some more notes to this section in https://github.com/ChainSafe/gossamer/pull/4519/commits/d7bd44cb3c06f9cc33583913c0ed8b139fe5750e